### ISSUE TITLE
Disallow reflection setting readonly static fields

### ISSUE THREAD

**Issue by MichalStrehovsky** [dotnet/runtimelab#200]
reflection/SetValue/TrySetReadonlyStaticField
reflection/SetValue/TrySetReadonlyStaticField2

**Comment by MichalStrehovsky** [comment]
Add `IsInitOnly` flag to the FieldTableFlags enum:

https://github.com/dotnet/runtimelab/blob/beef8b0fad689f8d7a7f7b2ec0f72263698e658e/src/coreclr/src/tools/Common/Internal/Runtime/MappingTableFlags.cs#L40-L54

Next, emit the flag from the AOT compiler - we're computing the value for the flags around here:

https://github.com/dotnet/runtimelab/blob/beef8b0fad689f8d7a7f7b2ec0f72263698e658e/src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReflectionFieldMapNode.cs#L86-L93

Add a line that sets the flag based on `field.IsInitOnly`.

Now we just need to respect that flag at runtime. The code that interprets it is here:

https://github.com/dotnet/runtimelab/blob/beef8b0fad689f8d7a7f7b2ec0f72263698e658e/src/coreclr/src/nativeaot/System.Private.Reflection.Execution/src/Internal/Reflection/Execution/ExecutionEnvironmentImplementation.MappingTables.cs#L1376-L1380

Here we're passing `fieldBase` to the accessors. `fieldBase` is a stripped version of `fieldAccessMetadata.Flags` - change the code to pass `fieldAccessMetadata.Flags` instead so that the accessors get the full flags, not just the lowest two bits. Note that the code expect the rest of the flags to be dropped, so check for places that do `==` with these values - they'll have to do the `& FieldTableFlags.StorageClass ` mask first

Then update the `StaticFieldAccessor` class to throw if we're trying to set a field that has the IsInitOnly flag set.

**Comment by rohithv1997** [comment]
@MichalStrehovsky - I would like to take this up if no one else is assigned on this

**Comment by MichalStrehovsky** [comment]
@rohithv1997 assigned to you! Thank you!

**Comment by rohithv1997** [comment]
Hi @MichalStrehovsky ,
Can you let me know how to change the status of this bug?

**Comment by MichalStrehovsky** [comment]
> Can you let me know how to change the status of this bug?

We had to leave one test blocked on this issue because we should apparently allow SetValue to succeed if it's being run from the static constructor itself. So we either keep this issue open to track that, or file a new issue and update issues.targets to reference the new issue. I don't have a preference. I just don't want issues.targets to refer to a closed issue.
